File: Recursion/Leetcode/Medium/CombinationSum_III/Combination_Sum_III.java
// https://leetcode.com/problems/combination-sum-iii/

import java.util.ArrayList;
import java.util.List;

public class Combination_Sum_III {
    static void generate(int k, int n, int curr, List<Integer> seq, List<List<Integer>> combo) {
        if(n <= 0 || curr >= 10 || k <= 0) {
            if(n == 0 && k == 0) {
                combo.add(new ArrayList<>(seq));
            }
            return;
        }
        if(n >= curr) {
            seq.add(curr);
            generate(k-1, n-curr, curr+1, seq, combo);
            seq.remove(seq.size() - 1);
            generate(k, n, curr+1, seq, combo);
        }
    }
    public static List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> list = new ArrayList<>();
        List<Integer> seq = new ArrayList<>();
        generate(k, n, 1, seq, list);
        return list;
    }

    public static void main(String[] args) {
        System.out.println(combinationSum3(3, 7));
    }
}

File: Recursion/Leetcode/Medium/CombinationSum_III/CombinationSum_III.cpp
// https://leetcode.com/problems/combination-sum-iii/
#include<bits/stdc++.h>
using namespace std;
void findCombo(int k, int n, vector<vector<int>> &combo, vector<int> &seq, int current)
{
    if (n == 0 || current == 10)
    {
        if (seq.size() == k && n == 0)
        {
            combo.push_back(seq);
        }
        return;
    }
    if (n >= current)
    {
        seq.push_back(current);
        findCombo(k, n - current, combo, seq, current + 1);
        seq.pop_back();
        findCombo(k, n, combo, seq, current + 1);
    }
}
vector<vector<int>> combinationSum3(int k, int n)
{
    vector<vector<int>> combo;
    vector<int> seq;
    findCombo(k, n, combo, seq, 1);
    return combo;
}
int main(void) {
    vector<vector<int>> ans = combinationSum3(3, 7);
    for(auto v: ans ){
        for(int ele: v) {
            cout << ele << " ";
        }
        cout << endl;
    }

    return 0;
}